src/tasks/util/advisory_lock.rs

```diff
@@ -0,0 +1,102 @@
+use std::error::Error;
+
+use diesel::prelude::*;
+use diesel::sql_types::BigInt;
+use diesel::PgConnection;
+
+sql_function!(fn pg_try_advisory_lock(key: BigInt) -> Bool);
+sql_function!(fn pg_advisory_unlock(key: BigInt) -> Bool);
+
+/// Run the callback if the session advisory lock for the given key can be obtained.
+///
+/// If the lock is not already held, the callback will be called and the lock will be unlocked
+/// after the closure returns.  If the lock is already held, the function returns an error without
+/// calling the callback.
+pub(crate) fn with_advisory_lock<F>(
+    conn: &PgConnection,
+    key: i64,
+    f: F,
+) -> Result<(), Box<dyn Error>>
+where
+    F: FnOnce(&PgConnection) -> QueryResult<()>,
+{
+    if !diesel::select(pg_try_advisory_lock(key)).get_result(conn)? {
+        let string = format!(
+            "A job holding the session advisory lock for key {} is already running",
+            key
+        );
+        println!("return");
+        return Err(string.into());
+    }
+    println!("umm");
+    let _dont_drop_yet = DropGuard { conn, key };
+    f(conn).map_err(Into::into)
+}
+
+struct DropGuard<'a> {
+    conn: &'a PgConnection,
+    key: i64,
+}
+
+impl<'a> Drop for DropGuard<'a> {
+    fn drop(&mut self) {
+        match diesel::select(pg_advisory_unlock(self.key)).get_result(self.conn) {
+            Ok(true) => (),
+            Ok(false) => println!(
+                "Error: job advisory lock for key {} was not locked",
```

Comment:
Maybe this should be a panic, to forbid the callback from releasing the lock on its own.